Back identifier tokens in custom properties with string literals

anttijk opened the pull request:
#### a24a25b87e86c3c62d9e4ca0cc1e8f4d3939c8eb
<pre>
Back identifier tokens in custom properties with string literals
<a href="https://bugs.webkit.org/show_bug.cgi?id=269986">https://bugs.webkit.org/show_bug.cgi?id=269986</a>
<a href="https://rdar.apple.com/123506269">rdar://123506269</a>

Reviewed by Cameron McCormack.

Tokens don&apos;t own the underlying string so we construct a separate backing string for tokens that need it.
Identifier tokens can in mosts cases be backed by their string literals, saving memory and work.

* Source/WebCore/css/CSSVariableData.cpp:
(WebCore::CSSVariableData::updateBackingStringsInTokens):
(WebCore::CSSVariableData::CSSVariableData):

Point the string backing to a literal if possible.

* Source/WebCore/css/parser/CSSParserToken.cpp:
(WebCore::CSSParserToken::id const):
(WebCore::CSSParserToken::functionId const):
(WebCore::CSSParserToken::identOrFunctionId const):
(WebCore::CSSParserToken::tryUseStringLiteralBacking):

Known identifier tokens can use literal backing if cases match (usually all lowercase).

(WebCore::CSSParserToken::copyWithUpdatedString const): Deleted.

Remove unsed function.

* Source/WebCore/css/parser/CSSParserToken.h:
(WebCore::CSSParserToken::isBackedByStringLiteral const):
(WebCore::CSSParserToken::initValueFromStringView):

* Source/WebCore/css/parser/CSSParserToken.h:

Canonical link: <a href="https://commits.webkit.org/275375@main">https://commits.webkit.org/275375@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7571606c73095e8088e4afd15cbbd84fc9b029bd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41695 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20709 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44076 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44267 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37788 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44002 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23836 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18039 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34455 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42269 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17632 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35906 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15127 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/15325 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36927 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45654 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37876 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37249 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/40995 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16510 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13550 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39420 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18129 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9339 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18186 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17773 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/css/CSSVariableData.cpp:
```diff
@@ -60,14 +60,18 @@ bool CSSVariableData::operator==(const CSSVariableData& other) const
+        if (token.tryUseStringLiteralBacking())
+            continue;
+        stringBuilder.append(token.value());
+    };
```
Nit: no `;`

heycam commented on Source/WebCore/css/parser/CSSParserToken.cpp:
```diff
@@ -486,15 +486,19 @@ CSSValueID CSSParserToken::id() const
+    return anyId();
+}
+
+CSSValueID CSSParserToken::anyId() const
```
Nit: I find the name `anyId` a little confusing. It sounds like it could return an arbitrary `CSSValueID`. What about `identOrFunctionId`?

anttijk committed:
Back identifier tokens in custom properties with string literals
https://bugs.webkit.org/show_bug.cgi?id=269986
rdar://123506269

Reviewed by Cameron McCormack.

Tokens don't own the underlying string so we construct a separate backing string for tokens that need it.
Identifier tokens can in mosts cases be backed by their string literals, saving memory and work.

* Source/WebCore/css/CSSVariableData.cpp:
(WebCore::CSSVariableData::updateBackingStringsInTokens):
(WebCore::CSSVariableData::CSSVariableData):

Point the string backing to a literal if possible.

* Source/WebCore/css/parser/CSSParserToken.cpp:
(WebCore::CSSParserToken::id const):
(WebCore::CSSParserToken::functionId const):
(WebCore::CSSParserToken::identOrFunctionId const):
(WebCore::CSSParserToken::tryUseStringLiteralBacking):

Known identifier tokens can use literal backing if cases match (usually all lowercase).

(WebCore::CSSParserToken::copyWithUpdatedString const): Deleted.

Remove unsed function.

* Source/WebCore/css/parser/CSSParserToken.h:
(WebCore::CSSParserToken::isBackedByStringLiteral const):
(WebCore::CSSParserToken::initValueFromStringView):

* Source/WebCore/css/parser/CSSParserToken.h:

Canonical link: https://commits.webkit.org/275375@main
